### PR TITLE
Add doPrivilege blocks for socket connect ops in repository-hdfs

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #22116. The repository-hdfs plugin opens socket
connections. As `SocketPermission` is transitioned out of core, hdfs
will require `connect` permission. This pull request wraps operations
that require this permission in `doPrivileged` blocks.

### CODE REVIEW

**Comment by tbrooks8** [comment]
In hdfs we were [calling](https://github.com/elastic/elasticsearch/pull/22793/files#diff-0ec2356265a4ebf55b306a8b6d18391aR121) doPrivileged and explicitly defining the permissions. I was not sure if there was a reason we were doing this differently, so I left it as is and just added `new SocketPermission("*", "connect")`.

**Comment by rjernst** [comment]
> In hdfs we are calling doPrivileged and explicitly defining the permissions

We do this to limit what craziness hdfs can do (and know exactly what it requires).

**Review by rjernst — APPROVED** [review]
LGTM

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java [hunk 1]
```diff
@@ -98,11 +90,20 @@ public InputStream readBlob(String blobName) throws IOException {
...
-            }
+        return store.execute((Operation<InputStream>) fileContext -> {
+            FSDataInputStream is = fileContext.open(new Path(path, blobName), bufferSize);
+            return new InputStream() {
+                @Override
+                public int read() throws IOException {
+                    try {
+                        SpecialPermission.check();
+                        // FSDataInputStream can open connection on read()
+                        return AccessController.doPrivileged((PrivilegedExceptionAction<Integer>) is::read);
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java]: can we maybe try to trigger this differently? I mean can we for instance try to call `#available()` or can we maybe read the first byte on open and wrap in a `BufferedInputStream` and then do this:

```Java

InputStream stream = is.isMarkSupported() ? is : new BufferedInputStream(is);
// do the following in doPrivileged?
stream.mark(1);
stream.skip(1);
stream.reset();
return stream;
```

**Review by s1monw — COMMENTED** [review]
I left one comment looks great so far

**Comment by tbrooks8** [comment]
@s1monw In my testing available() was not doing the trick. So I went with the buffered, mark, skip, reset approach.

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java [hunk 2]
```diff
@@ -90,20 +91,20 @@ public InputStream readBlob(String blobName) throws IOException {
...
-                    try {
-                        SpecialPermission.check();
-                        // FSDataInputStream can open connection on read()
-                        return AccessController.doPrivileged((PrivilegedExceptionAction<Integer>) is::read);
-                    } catch (PrivilegedActionException e) {
-                        throw (IOException) e.getCause();
-                    }
-                }
-            };
+            InputStream stream = is.markSupported() ? is : new BufferedInputStream(is);
```
Comment (s1monw) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java]: if we run into an exception here we have to close the stream. 

we usually do this:

```Java
boolean success = false;
try {
 // do something with the stream
  success = true;
  return stream;
} finally {
  if (success == false) {
    IOUtils.closeWhileHandlingException(stream);
  }
}

**Review by s1monw — CHANGES_REQUESTED** [review]
left one more comment we are close I think

**Comment by tbrooks8** [comment]
Note to other reviews:

Simon and I talked about this for a while and decided it was safest to return a `InputStream` implementation that wraps all of the methods.

**Review by s1monw — APPROVED** [review]
LGTM

**Pull request merged**